File: internal/messaging/slack/backoff.go
package slack

import (
	"math/rand/v2"
	"sync"
	"time"
)

// reconnectBackoff implements exponential backoff with jitter for reconnection attempts.
// It is safe for concurrent use.
type reconnectBackoff struct {
	attempt   int
	baseDelay time.Duration
	maxDelay  time.Duration
	mu        sync.Mutex
}

// newReconnectBackoff creates a new backoff with the specified base and max delays.
func newReconnectBackoff(baseDelay, maxDelay time.Duration) *reconnectBackoff {
	return &reconnectBackoff{
		attempt:   0,
		baseDelay: baseDelay,
		maxDelay:  maxDelay,
	}
}

// Next returns the next backoff duration with exponential backoff and jitter.
// Algorithm:
//   - delay = baseDelay * 2^attempt
//   - if delay > maxDelay → delay = maxDelay
//   - jitter = random duration in [0, delay/2)
//   - final = delay/2 + jitter (ensures at least half the computed delay)
//   - increment attempt
func (b *reconnectBackoff) Next() time.Duration {
	b.mu.Lock()
	defer b.mu.Unlock()

	delay := b.baseDelay << b.attempt
	if delay <= 0 || delay > b.maxDelay {
		delay = b.maxDelay
	}

	halfDelay := delay / 2
	if halfDelay > 0 {
		jitter := time.Duration(rand.Int64N(int64(halfDelay)))
		delay = halfDelay + jitter
	}

	b.attempt++
	if b.attempt > 30 {
		b.attempt = 30
	}
	return delay
}

// Reset resets the backoff attempt counter to 0.
func (b *reconnectBackoff) Reset() {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.attempt = 0
}

File: examples/typescript-client/src/backoff.ts
/**
 * Jittered Exponential Backoff implementation
 */

export interface BackoffConfig {
  baseDelayMs: number;
  maxDelayMs: number;
  jitter?: number; // 0 to 1, default 0.1
}

export function calculateBackoff(attempt: number, config: BackoffConfig): number {
  if (attempt <= 0) return 0;
  
  const { baseDelayMs, maxDelayMs, jitter = 0.1 } = config;
  
  // Exponential backoff: base * 2^(attempt-1)
  let delay = baseDelayMs * Math.pow(2, attempt - 1);
  
  // Cap at maxDelay
  delay = Math.min(delay, maxDelayMs);
  
  // Add jitter
  if (jitter > 0) {
    const jitterAmount = delay * jitter;
    const randomJitter = (Math.random() * 2 - 1) * jitterAmount;
    delay += randomJitter;
  }
  
  return Math.max(0, delay);
}
